feat(websocket): enhance reconnection logic with intelligent error classification

learnerLj opened the pull request:
## Summary

This PR enhances the WebSocket reconnection logic with intelligent error classification to provide more reliable connections for financial applications.

## Key Improvements

### 🔧 **Smart Error Classification**
- Introduced `ConnectionFailureReason` enum to classify WebSocket failures
- Distinguishes between **recoverable errors** (network issues, temporary server errors) and **permanent failures** (authentication, protocol violations)
- Follows RFC 6455 WebSocket close code specifications

### 🚀 **Enhanced Reconnection Coverage**
- **Previously**: Only handled `Message::Close` with non-normal codes
- **Now**: Handles network errors (`Err(e)`), stream ending, and writer task failures
- Prevents unnecessary reconnection attempts for permanent failures

### 🛡️ **Financial-Grade Reliability**
- Avoids reconnection for authentication failures and protocol violations
- Provides detailed error classification for monitoring and debugging
- Maintains backward compatibility with all existing functionality

## Technical Details

### Error Classification Logic

**Reconnectable Errors:**
- Network interruptions and connection resets
- Temporary server errors (codes 1001, 1011-1013)
- Unexpected closes (code 1006) and stream endings

**Non-Reconnectable Errors:**
- Authentication failures and permanent server errors
- Protocol violations (codes 1002, 1003, 1007, 1009, 1010)
- Configuration errors and user-initiated closes
- Normal closure (code 1000)

### New Features
1. **Writer Task Error Handling**: Writer failures now trigger reconnection for recoverable errors
2. **Stream End Detection**: Handles unexpected stream termination scenarios
3. **Enhanced Logging**: Detailed error classification for better observability

## Testing

- ✅ All 162 existing WebSocket tests pass
- ✅ Verified with production WebSocket stream examples
- ✅ No regressions or breaking changes
- ✅ Maintains full backward compatibility

## Use Case

This enhancement is particularly valuable for:
- Financial trading applications requiring high reliability
- Long-running WebSocket connections
- Applications needing intelligent reconnection strategies
- Production systems requiring detailed error monitoring

## Examples

### Before (Original Logic)
```rust
// Only checked if close code was normal
CloseCode::from(code) \!= CloseCode::Normal
```

### After (Enhanced Logic)
```rust
// Intelligent classification with detailed error handling
let failure_reason = ConnectionFailureReason::from_close_code(code, user_initiated);
if failure_reason.should_reconnect() {
    // Smart reconnection decision
}
```

This change significantly improves WebSocket connection reliability while maintaining the existing API contract.

dimitrisn442 commented:
Hi @learnerLj, your PR looks great. Thanks for contributing! A couple of comments:

- Could you please move the `ConnectionFailureReason` on the `errors` file (and maybe rename it to `WebsocketConnectionFailureReason`?
- Can you write some tests for the new error handling logic?
- Could you please also run the linter before pushing? We use `clippy` for our linting (`cargo clippy --fix --allow-dirty --allow-staged --all-targets`

learnerLj commented:
Hi @dimitrisn442, thank you for the review and feedback\! I've addressed all your comments:

✅ **Moved and renamed the enum**: 
- Moved `ConnectionFailureReason` from `websocket.rs` to `errors.rs`
- Renamed to `WebsocketConnectionFailureReason` for better clarity
- Updated all imports and usages throughout the codebase

✅ **Added comprehensive tests**:
- Created 8 test functions covering all error handling logic
- Tests include `from_tungstenite_error()` classification for various error types
- Tests cover `from_close_code()` with standard WebSocket close codes
- Tests validate `should_reconnect()` logic for all error variants
- All tests pass successfully ✅

✅ **Ran clippy linter**:
- Executed `cargo clippy --fix --allow-dirty --allow-staged --all-targets`
- Fixed all match arm duplication warnings (consolidated 28 warnings → 0)
- Code now follows Rust best practices and clippy recommendations
- Only remaining warning is a harmless resolver configuration notice

## Changes Summary
- **Better code organization**: Error handling logic now properly separated
- **Comprehensive test coverage**: All error classification scenarios tested
- **Clean, lint-free code**: Follows Rust coding standards
- **Maintained functionality**: All existing tests pass, no breaking changes

The PR is now ready for another review. Thanks for helping improve the code quality\!

dimitrisn442 commented on src/common/errors.rs:
```diff
@@ -1,5 +1,93 @@
 use thiserror::Error;
 
+/// Represents different types of WebSocket connection failures and their reconnection eligibility
+#[derive(Debug, Clone)]
```
this has been addressed
